### ISSUE TITLE
namespaceSelector use the same scheme 'NamespaceSelector'

### ISSUE THREAD

**Issue by QCU266** [prometheus-operator/prometheus-operator#3553]
**What is missing?**
* PrometheusSpec.serviceMonitorNamespaceSelector  using Scheme [metav1.LabelSelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#labelselector-v1-meta), 
but MonitorSpec.namespaceSelector using scheme [NamespaceSelector](https://github.com/prometheus-operator/prometheus-operator/blob/release-0.38/Documentation/api.md#namespaceselector)

**Why do we need it?**
This inconsistency can easily lead to confusion and configuration errors. 
Why not Why not  all **namespaceSelector** use the same scheme(e.g. [NamespaceSelector](https://github.com/prometheus-operator/prometheus-operator/blob/release-0.38/Documentation/api.md#namespaceselector))?

**Environment**

* Prometheus Operator version:
    `0.42`

**Anything else we need to know?**:

**Comment by brancz** [comment]
I agree that this is confusing, and it has indeed lead to many issues on this repo. The problem is that Prometheus' namespace selection mechanism is not label based (it can only take a list of namespaces or all namespaces). In hindsight, we should have not called this field `namespaceSelector` but probably just `namespaces`. We could certainly try and improve this by renaming. What do you think, would that actually be clearer to people?

**Comment by coderanger** [comment]
@brancz Another option would be to have the operator do the label selector and pass in a list of namespaces?

**Comment by lilic** [comment]
I agreed, with @brancz approach, but I believe a lot of people already heavily use this, so I would prefer to add a new field if its causing this much problems rather than renaming current one and we just drop docs for the old one.

cc @prometheus-operator/prometheus-operator-triage WDYT?

**Comment by paulfantom** [comment]
renaming the current one and dropping `namespaceSelector` from docs while still having it as a valid option SGTM.

**Comment by s-urbaniak** [comment]
> add a new field if its causing this much problems rather than renaming current one and we just drop docs for the old one.
I like this approach a lot :+1:

**Comment by nrchakradhar** [comment]
![PrometheusOperatorSelector-For GIT](https://user-images.githubusercontent.com/38882990/97586735-9cfc7b00-1a20-11eb-86fd-a7980f6d79d1.png)
I made a diagram for my own understanding. I am still not clear as to which fields will be kept, renamed and what scenarios will be covered. 

I am not able to understand 
> The problem is that Prometheus' namespace selection mechanism is not label based (it can only take a list of namespaces or all namespaces).

Also about the TODO in the code:
> https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/apis/monitoring/v1/types.go#L1331-L1333

What are the desired changes and what scenarios should be covered. If some small description is shared, I can make a writeup for the same.

My current understanding is as below:
Prometheus can run in one namespace and scrape targets can be in multiple different namespaces. 
Scrape targets are defined by ServiceMonitor, PodMonitor, Probe CR created in different namespaces. 
Multiple Prometheuses in different namespaces can scrape the same target.
Prometheus operator generates the prometheus.yaml for each prometheus instance deployed. The operator picks the targets based on the configuration of *selectors in kind:Prometheus that match the labels for CRs in selected namespaces.

The ServiceMonitor, PodMonitor, Probe CR select (rather point) to the actual targets through similar selection mechanism.

**Comment by s-urbaniak** [comment]
@nrchakradhar let me think about your comment a bit as I believe you have some valid question, but first and foremost I would like to say a big thank you!!!!!! from me and the @prometheus-operator/prometheus-operator-reviewers team for the picture you drew and the effort, that helps a lot in thinking about it more deeply :tada: :heart:

**Comment by nrchakradhar** [comment]
@s-urbaniak For having a better discussion that I can understand, I have created another diagram. As per my understanding all the 3 scenarios indicated are possible currently. @prometheus-operator/prometheus-operator-reviewers, please let me know if any of them is not possible to be achieved. Based on the discussions here, example yaml's can be created for benefit of all.

![PrometheusOperatorSelector-ForGitWithNamespaces](https://user-images.githubusercontent.com/38882990/97805307-49d23480-1c7b-11eb-8bb5-26b7098bf8fb.png)

**Comment by nrchakradhar** [comment]
![PrometheusOperatorSelector-ForGitWithNamespaces](https://user-images.githubusercontent.com/38882990/99071346-fb585a80-25d7-11eb-862c-1658589aabf3.png)

I hope, I have covered most possible use cases in the picture above. I would be happy to receive any feedback/corrections/suggestions.

**Comment by paulfantom** [comment]
Help needed to move those awesome graphs into project documentation!

**Comment by slashpai** [comment]
https://github.com/prometheus-operator/prometheus-operator/issues/3553#issuecomment-726733177 this diagram?

**Comment by paulfantom** [comment]
@slashpai correct

**Comment by SerikiAyodele** [comment]
Considering picking this up, @paulfantom what part of the documentation should the diagram be added to?🤔

**Comment by ArthurSens** [comment]
Before adding such thing, I think we'll need to revisit how we organize our documentation (that's why we added the project idea to GSoC).

If we're adding graphs, I think we might want to start adding a basic architecture to the [introduction](https://prometheus-operator.dev/docs/prologue/introduction/) page first. Namespace selection is a bit more advanced and we don't have anything talking about this in our docs, so there's nowhere to put it right now 😬
